File: src/main/java/org/taktik/icure/logic/impl/HealthElementLogicImpl.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.logic.impl;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.taktik.icure.dao.HealthElementDAO;
import org.taktik.icure.dao.Option;
import org.taktik.icure.dao.impl.idgenerators.UUIDGenerator;
import org.taktik.icure.entities.HealthElement;
import org.taktik.icure.entities.Patient;
import org.taktik.icure.entities.embed.Delegation;
import org.taktik.icure.logic.HealthElementLogic;
import org.taktik.icure.logic.ICureSessionLogic;
import org.taktik.icure.utils.FuzzyValues;
import org.taktik.icure.validation.aspect.Check;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import javax.validation.constraints.NotNull;


/**
 * Created by emad7105 on 24/06/2014.
 */
@org.springframework.stereotype.Service
public class HealthElementLogicImpl extends GenericLogicImpl<HealthElement, HealthElementDAO> implements HealthElementLogic {
	private static final Logger log = LoggerFactory.getLogger(HealthElementLogicImpl.class);


	private HealthElementDAO healthElementDAO;
	private UUIDGenerator uuidGenerator;
	private ICureSessionLogic sessionLogic;

	@Autowired
	public void setHealthElementDAO(HealthElementDAO healthElementDAO) {
		this.healthElementDAO = healthElementDAO;
	}
	@Autowired
	public void setUuidGenerator(UUIDGenerator uuidGenerator) {
		this.uuidGenerator = uuidGenerator;
	}
	@Autowired
	public void setSessionLogic(ICureSessionLogic sessionLogic) {
		this.sessionLogic = sessionLogic;
	}
	@Override
	protected HealthElementDAO getGenericDAO() {
		return healthElementDAO;
	}

	@Override
	public HealthElement createHealthElement(@Check @NotNull HealthElement healthElement) {
		List<HealthElement> createdHealthElements = new ArrayList<>(1);
		try {
			// Fetching the hcParty
			String healthcarePartyId = sessionLogic.getCurrentHealthcarePartyId();

			// Setting Healthcare problem attributes
			healthElement.setId(uuidGenerator.newGUID().toString());
			if (healthElement.getOpeningDate() == null) { healthElement.setOpeningDate(FuzzyValues.getFuzzyDateTime(LocalDateTime.now(), ChronoUnit.SECONDS)); }
			healthElement.setAuthor(healthcarePartyId);
			healthElement.setResponsible(healthcarePartyId);

			// TODO should we check that opening contacts or closing contacts are valid?

			createEntities(Collections.singleton(healthElement), createdHealthElements);
		} catch (Exception e) {
			log.error("createHealthElement: " + e.getMessage());
			throw new IllegalArgumentException("Invalid Healthcare problem", e);
		}
		return createdHealthElements.size() == 0 ? null:createdHealthElements.get(0);
	}

	@Override
	public HealthElement getHealthElement(String healthElementId) {
		return healthElementDAO.getHealthElement(healthElementId);
	}

	@Override
	public List<HealthElement> getHealthElements(List<String> healthElementIds) {
		return healthElementDAO.getList(healthElementIds);
	}

	@Override
	public List<HealthElement> findByHCPartySecretPatientKeys(String hcPartyId, List<String> secretPatientKeys) {
		return healthElementDAO.findByHCPartySecretPatientKeys(hcPartyId, secretPatientKeys);
	}

	@Override
	public Set<String> deleteHealthElements(Set<String> ids) {
		try {
			deleteEntities(ids);
		} catch (Exception e) {
			log.error(e.getMessage(), e);
			return null;
		}
		return ids;
	}

	@Override
	public HealthElement modifyHealthElement(@Check @NotNull HealthElement healthElement) {
		try {
			updateEntities(Collections.singleton(healthElement));
            return getHealthElement(healthElement.getId());
		} catch (Exception e) {
			throw new IllegalArgumentException("Invalid Health problem", e);
		}
	}

	@Override
	public HealthElement addDelegation(String healthElementId, String healthcarePartyId, Delegation delegation) {
		HealthElement healthElement = getHealthElement(healthElementId);
		healthElement.addDelegation(healthcarePartyId, delegation);

		return healthElementDAO.save(healthElement);
	}

	@Override
	public HealthElement addDelegations(String healthElementId, List<Delegation> delegations) {
		HealthElement healthElement = getHealthElement(healthElementId);
		delegations.forEach(d->healthElement.addDelegation(d.getDelegatedTo(),d));

		return healthElementDAO.save(healthElement);

	}

	@Override
	public void solveConflicts() {
		List<HealthElement> healthElementsInConflict = healthElementDAO.listConflicts().stream().map(it -> healthElementDAO.get(it.getId(), Option.CONFLICTS)).collect(Collectors.toList());
		healthElementsInConflict.forEach(p-> {
			Arrays.stream(p.getConflicts()).map(c->healthElementDAO.get(p.getId(),c)).forEach(cp -> {
				p.solveConflictWith(cp);
				healthElementDAO.purge(cp);
			});
			healthElementDAO.save(p);
		});

	}


}

File: src/main/kotlin/org/taktik/icure/asynclogic/impl/HealthElementLogicImpl.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.asynclogic.impl

import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.dropWhile
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.mapNotNull
import kotlinx.coroutines.flow.take
import kotlinx.coroutines.flow.toList
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.taktik.couchdb.DocIdentifier
import org.taktik.couchdb.ViewQueryResultEvent
import org.taktik.couchdb.entity.Option
import org.taktik.couchdb.id.UUIDGenerator
import org.taktik.icure.asyncdao.HealthElementDAO
import org.taktik.icure.asynclogic.AsyncSessionLogic
import org.taktik.icure.asynclogic.HealthElementLogic
import org.taktik.icure.asynclogic.impl.filter.Filters
import org.taktik.icure.db.PaginationOffset
import org.taktik.icure.domain.filter.chain.FilterChain
import org.taktik.icure.entities.HealthElement
import org.taktik.icure.entities.embed.Delegation
import org.taktik.icure.entities.embed.Identifier

/**
 * Created by emad7105 on 24/06/2014.
 */
@ExperimentalCoroutinesApi
@Service
class HealthElementLogicImpl(private val filters: Filters,
                             private val healthElementDAO: HealthElementDAO,
                             private val uuidGenerator: UUIDGenerator,
                             private val sessionLogic: AsyncSessionLogic) : GenericLogicImpl<HealthElement, HealthElementDAO>(sessionLogic), HealthElementLogic {

    override fun getGenericDAO(): HealthElementDAO {
        return healthElementDAO
    }

    override suspend fun createHealthElement(healthElement: HealthElement) = fix(healthElement) { healthElement ->
        try { // Fetching the hcParty
            createEntities(setOf(healthElement)).firstOrNull()
        } catch (e: Exception) {
            log.error("createHealthElement: " + e.message)
            throw IllegalArgumentException("Invalid Healthcare problem", e)
        }
    }

    override suspend fun getHealthElement(healthElementId: String): HealthElement? {
        return healthElementDAO.getHealthElement(healthElementId)
    }

    override fun getHealthElements(healthElementIds: List<String>): Flow<HealthElement> = flow {
        emitAll(healthElementDAO.getEntities(healthElementIds))
    }

    override fun listHealthElementsByHcPartyAndSecretPatientKeys(hcPartyId: String, secretPatientKeys: List<String>) = flow {
        emitAll(healthElementDAO.listHealthElementsByHCPartyAndSecretPatientKeys(hcPartyId, secretPatientKeys))
    }

    override fun listHealthElementIdsByHcParty(hcpId: String) = flow {
        emitAll(healthElementDAO.listHealthElementsByHcParty(hcpId))
    }

    override fun listHealthElementIdsByHcPartyAndSecretPatientKeys(hcPartyId: String, secretPatinetKeys: List<String>) = flow {
        emitAll(healthElementDAO.listHealthElementIdsByHcPartyAndSecretPatientKeys(hcPartyId, secretPatinetKeys))
    }

    override suspend fun listLatestHealthElementsByHcPartyAndSecretPatientKeys(hcPartyId: String, secretPatientKeys: List<String>): List<HealthElement> {
        return healthElementDAO.listHealthElementsByHCPartyAndSecretPatientKeys(hcPartyId, secretPatientKeys).toList()
                .groupBy { it.healthElementId }.values.mapNotNull { value -> value.maxByOrNull { it: HealthElement ->
                    it.modified ?: it.created ?: 0L
                } }
    }

    override fun listHealthElementIdsByHcPartyAndCodes(hcPartyId: String, codeType: String, codeNumber: String) = flow {
        emitAll(healthElementDAO.listHealthElementsByHcPartyAndCodes(hcPartyId, codeType, codeNumber))
    }

    override fun listHealthElementIdsByHcPartyAndTags(hcPartyId: String, tagType: String, tagCode: String) = flow {
        emitAll(healthElementDAO.listHealthElementsByHcPartyAndTags(hcPartyId, tagType, tagCode))
    }

    override fun listHealthElementsIdsByHcPartyAndIdentifiers(hcPartyId: String, identifiers: List<Identifier>) = flow {
        emitAll(healthElementDAO.listHealthElementsIdsByHcPartyAndIdentifiers(hcPartyId, identifiers))
    }

    override fun listHealthElementIdsByHcPartyAndStatus(hcPartyId: String, status: Int) = flow {
        emitAll(healthElementDAO.listHealthElementsByHcPartyAndStatus(hcPartyId, status))
    }

    override fun deleteHealthElements(ids: Set<String>): Flow<DocIdentifier> {
        return try {
            deleteEntities(ids)
        } catch (e: Exception) {
            log.error(e.message, e)
            flowOf()
        }
    }

    override suspend fun modifyHealthElement(healthElement: HealthElement) = fix(healthElement) { healthElement ->
        try {
            modifyEntities(setOf(healthElement)).firstOrNull()
        } catch (e: Exception) {
            throw IllegalArgumentException("Invalid Health problem", e)
        }
    }

    override suspend fun addDelegation(healthElementId: String, delegation: Delegation): HealthElement? {
        val healthElement = getHealthElement(healthElementId)
        return delegation.delegatedTo?.let { healthcarePartyId ->
            healthElement?.let { c -> healthElementDAO.save(c.copy(delegations = c.delegations + mapOf(
                    healthcarePartyId to setOf(delegation)
            )))}
        } ?: healthElement
    }

    override suspend fun addDelegations(healthElementId: String, delegations: List<Delegation>): HealthElement? {
        val healthElement = getHealthElement(healthElementId)
        return healthElement?.let {
            return healthElementDAO.save(it.copy(
                    delegations = it.delegations +
                            delegations.mapNotNull { d -> d.delegatedTo?.let { delegateTo -> delegateTo to setOf(d) } }
            ))
        }
    }

    override fun solveConflicts(): Flow<HealthElement> =
        healthElementDAO.listConflicts().mapNotNull {
            healthElementDAO.get(it.id, Option.CONFLICTS)?.let { healthElement ->
                healthElement.conflicts?.mapNotNull { conflictingRevision ->
                    healthElementDAO.get(
                        healthElement.id,
                        conflictingRevision
                    )
                }
                    ?.fold(healthElement) { kept, conflict ->
                        kept.merge(conflict).also { healthElementDAO.purge(conflict) }
                    }
                    ?.let { mergedHealthElement -> healthElementDAO.save(mergedHealthElement) }
            }
        }


    override fun filter(paginationOffset: PaginationOffset<Nothing>, filter: FilterChain<HealthElement>) =
        flow<ViewQueryResultEvent> {
            val ids = filters.resolve(filter.filter)
            val sortedIds = paginationOffset.takeUnless { it.startDocumentId == null }
                ?.let { paginationOffset -> // Sub-set starting from startDocId to the end (including last element)
                    ids.dropWhile { id -> id != paginationOffset.startDocumentId }
                } ?: ids

            val selectedIds =
                sortedIds.take(paginationOffset.limit + 1) // Fetching one more health element for the start key of the next page
            emitAll(healthElementDAO.findHealthElementsByIds(selectedIds))
        }

    companion object {
        private val log = LoggerFactory.getLogger(HealthElementLogicImpl::class.java)
    }
}
